### PR TITLE
feat(runbooks): G12.2-T1 runbook template Pydantic schemas + step-shape validation

### LINKED ISSUE #1295: Task: G12.2-T1 Runbook Pydantic schemas + step-shape discriminated-union validation
Parent initiative: #1197
Parent goal: #1195

Depends-on: #1292

## Problem

Every G12.2 tool (draft / edit / publish / deprecate / list / show) and every G12.3 verb (start / next / abort / reassign / list-runs) takes or returns a `RunbookTemplate` or one of its sub-shapes. Without typed schemas, every route, MCP tool, and CLI verb would re-invent its own validation — and the step-shape contract (`type: operation_call | manual` with `verify: confirm | operation_call`) is the load-bearing data structure for adherence (G12.3).

This Task lands the Pydantic schemas module. **Pure types — no service logic, no routes, no tools.** Step-shape validation is the meat: discriminated unions for `step.type` and `verify.type`, plus a `${...}` allowlist that rejects every substitution pattern except `${run.target}` and `${run.params.X}` at publish time (defense in depth — G12.3 will re-check at advance time per its own scope).

## Current state

- [backend/src/meho_backplane/db/models.py](backend/src/meho_backplane/db/models.py) (after #1292 lands) will carry `RunbookTemplate`, `RunbookRun`, `RunbookRunStepState` as SQLAlchemy models — storage-level shapes only, no validation logic.
- Pydantic-schemas convention across the codebase: one `<module>/schemas.py` per feature area, every model `ConfigDict(frozen=True)`, request/response shapes paired with the route group. References:
  - [backend/src/meho_backplane/kb/schemas.py](backend/src/meho_backplane/kb/schemas.py) — `KbEntry`, `KbIngestionResult`, `SLUG_PATTERN` regex, `validate_slug()` helper. **Reuse `SLUG_PATTERN` (line 79) verbatim** for `runbook_templates.slug` validation.
  - [backend/src/meho_backplane/memory/schemas.py](backend/src/meho_backplane/memory/schemas.py) — five-scope discriminated shape with model-level validators (the closest precedent for runbook step-shape validation).
  - [backend/src/meho_backplane/scheduler/schemas.py:103](backend/src/meho_backplane/scheduler/schemas.py#L103) — `model_validator(mode="after")` enforcing "exactly-one-discriminator + per-kind field validation" with clean 422 surfacing for IDE-readable error messages. **This is the validation pattern** to mirror for `Step` / `Verify` discriminated unions.
- No `backend/src/meho_backplane/runbooks/` module exists yet. This Task creates it with `schemas.py` as the first file.

### Module structure

New package: `backend/src/meho_backplane/runbooks/__init__.py` (empty) + `backend/src/meho_backplane/runbooks/schemas.py`.

### Step shape — discriminated union

Two step types, both first-class per Initiative #1197 scope:

```python
class OperationCallStep(BaseModel):
    """Step the agent dispatches via ``call_operation()``."""
    model_config = ConfigDict(frozen=True)
    id: str  # template-author slug, validated against STEP_ID_PATTERN
    title: str
    body: str  # operator-readable instructions (Markdown; may contain ${run.target} / ${run.params.X})
    type: Literal["operation_call"]
    op_id: str  # the operation registry op id, e.g. "vmware.composite.vm.create"
    params: dict[str, object]  # may contain ${...} substitutions
    verify: Verify  # discriminated union, see below


class ManualStep(BaseModel):
    """Step the operator performs off-MEHO (SSH, web UI, etc)."""
    model_config = ConfigDict(frozen=True)
    id: str
    title: str
    body: str  # may contain ${run.target} / ${run.params.X}
    type: Literal["manual"]
    verify: Verify


Step = Annotated[
    OperationCallStep | ManualStep,
    Field(discriminator="type"),
]
```

### Verify shape — discriminated union

Two verify types per Initiative #1197 (verify is intentionally minimal — no JSONPath / no comparison operators / no boolean composition):

```python
class ConfirmVerify(BaseModel):
    """Operator answers yes/no/escalate; only ``yes`` advances."""
    model_config = ConfigDict(frozen=True)
    type: Literal["confirm"]
    prompt: str  # the question shown to the operator


class OperationCallVerify(BaseModel):
    """MEHO dispatches the verify call, structural-equality + presence match against expect."""
    model_config = ConfigDict(frozen=True)
    type: Literal["operation_call"]
    op_id: str
    params: dict[str, object]
    expect: dict[str, object]  # structural-equality + presence match (no operators)


Verify = Annotated[
    ConfirmVerify | OperationCallVerify,
    Field(discriminator="type"),
]
```

### Template shape

```python
class RunbookTemplateBody(BaseModel):
    """The author-facing template shape — what gets stored in
    ``runbook_templates.steps`` as JSONB."""
    model_config = ConfigDict(frozen=True)
    title: str
    description: str
    target_kind: str | None = None  # optional product hint
    steps: list[Step]  # ordered; verify gates advance; opacity at run time

    @model_validator(mode="after")
    def _validate_step_ids_unique_and_substitutions_allowlisted(self) -> RunbookTemplateBody:
        # 1. Step ids unique within the template.
        # 2. Every step.body / step.params / verify.params / verify.expect
        #    walked recursively to find every ${...} pattern; allow only
        #    ${run.target} and ${run.params.X} (X = [a-z_][a-z0-9_]*); any
        #    other pattern raises ValueError.
        ...
```

### Substitution allowlist

New helper `validate_substitutions(value: object) -> None` (module-level):
- Walks `value` recursively (str / dict / list).
- Finds every `${...}` pattern via regex.
- Rejects with `ValueError("disallowed substitution pattern: ${...}")` on any pattern except `${run.target}` and `${run.params.X}` where `X` matches `[a-z_][a-z0-9_]*`.
- Called from `_validate_step_ids_unique_and_substitutions_allowlisted` and re-exported on `__all__` for G12.3 to reuse at advance time (defense in depth).

### Step / slug id patterns

```python
#: Step id pattern — same shape as the kb slug pattern but allows
#: shorter identifiers. Step authors write ids like ``revoke-old-cert``,
#: ``rotate-credentials``, ``verify-cluster-quorum``.
STEP_ID_PATTERN: Final[re.Pattern[str]] = re.compile(r"^[a-z][a-z0-9\-]{0,63}$")
```

The `runbook_templates.slug` column reuses `kb.schemas.SLUG_PATTERN` verbatim (per #1292 / G12.1-T1 — already grounded). Import it; don't redefine.

### Tool request/response shapes

Six pairs, one per tool the G12.2 service layer (T2) will surface:

```python
class DraftTemplateRequest(BaseModel):
    slug: str  # validated against SLUG_PATTERN
    body: RunbookTemplateBody


class DraftTemplateResponse(BaseModel):
    slug: str
    version: int
    status: Literal["draft"]


class EditTemplateRequest(BaseModel):
    slug: str
    body: RunbookTemplateBody


class EditTemplateResponse(BaseModel):
    slug: str
    version: int  # same as input if editing a draft; new version if forking from a published template
    status: Literal["draft"]
    forked_from: ForkInfo | None = None  # populated on the fork path; null on draft-edit


class ForkInfo(BaseModel):
    """Surfaced by ``runbook_edit_template`` when editing a published template forks
    a new draft. The CLI / UI shows this so the senior knows what they're forking."""
    slug: str
    version: int  # the version being forked from
    in_flight_run_count: int  # how many runs are still pinned to the forked version


class PublishTemplateRequest(BaseModel):
    slug: str
    version: int


class PublishTemplateResponse(BaseModel):
    slug: str
    version: int
    status: Literal["published"]


class DeprecateTemplateRequest(BaseModel):
    slug: str
    version: int


class DeprecateTemplateResponse(BaseModel):
    slug: str
    version: int
    status: Literal["deprecated"]


class ListTemplatesFilter(BaseModel):
    status: Literal["draft", "published", "deprecated"] | None = None
    target_kind: str | None = None


class TemplateSummary(BaseModel):
    """Operator-readable summary surfaced by ``runbook_list_templates``."""
    slug: str
    version: int
    title: str
    status: Literal["draft", "published", "deprecated"]
    target_kind: str | None
    edited_at: datetime


class ShowTemplateResponse(BaseModel):
    """TENANT_ADMIN-readable full template surface."""
    slug: str
    version: int
    title: str
    description: str
    target_kind: str | None
    status: Literal["draft", "published", "deprecated"]
    steps: list[Step]
    created_by: str
    created_at: datetime
    edited_by: str
    edited_at: datetime
```

All models `ConfigDict(frozen=True)`. `__all__` populated alphabetically.

### Tests

New file [backend/tests/test_runbooks_schemas.py](backend/tests/test_runbooks_schemas.py):

1. **`test_operation_call_step_valid`** — minimal valid `OperationCallStep`.
2. **`test_manual_step_valid`** — minimal valid `ManualStep`.
3. **`test_step_unknown_type_rejected`** — `{"type": "magic", ...}` → `ValidationError`.
4. **`test_step_missing_id_rejected`** — `{"type": "manual", "body": "..."}` (no `id`) → `ValidationError`.
5. **`test_step_id_pattern_rejected`** — `id="Has-Caps"` → `ValidationError`.
6. **`test_confirm_verify_valid`** + **`test_operation_call_verify_valid`** — both verify shapes parse.
7. **`test_verify_unknown_type_rejected`** — discriminator catches `verify.type: "magic"`.
8. **`test_substitution_allowed`** — `${run.target}` and `${run.params.foo}` parse; surface them inside `step.body`, `step.params`, `verify.params`, `verify.expect`.
9. **`test_substitution_rejected`** — `${anything_else}`, `${run.bad}`, `${run.params.X.Y}` (no nested), `${run.params.WithCaps}` → `ValueError("disallowed substitution pattern: ...")`.
10. **`test_step_ids_must_be_unique`** — two steps with same `id` → `ValueError`.
11. **`test_fork_info_response_shape`** — `EditTemplateResponse` with `forked_from` populated; with `forked_from=None`; both parse.
12. **`test_list_templates_filter_optional_fields`** — `ListTemplatesFilter()` with no fields parses; filtering by `status`/`target_kind` parses.
13. **`test_template_body_roundtrip`** — build a `RunbookTemplateBody`, dump to dict, re-parse, assert equality (no information loss across the boundary that G12.2-T2's service writes into JSONB).
14. **`test_validate_substitutions_helper_exported`** — `from meho_backplane.runbooks.schemas import validate_substitutions` works; reused by G12.3.

## Acceptance criteria

- [ ] Package `backend/src/meho_backplane/runbooks/` exists with `__init__.py` and `schemas.py`.
- [ ] `Step` and `Verify` are `Annotated[... | ..., Field(discriminator="type")]` discriminated unions.
- [ ] `RunbookTemplateBody.steps` validates step ids unique within the template.
- [ ] `validate_substitutions()` rejects every `${...}` pattern except `${run.target}` and `${run.params.X}` (X matching `[a-z_][a-z0-9_]*`); helper is module-level and exported.
- [ ] `STEP_ID_PATTERN` defined; reused for step ids. `runbook_templates.slug` validation imports `SLUG_PATTERN` from `kb.schemas` (no redefinition).
- [ ] Every Pydantic model uses `ConfigDict(frozen=True)`.
- [ ] `__all__` populated alphabetically and complete.
- [ ] `uv run pytest backend/tests/test_runbooks_schemas.py -v` green (all 14 tests).
- [ ] `uv run ruff check backend/src/meho_backplane/runbooks/ && uv run mypy backend/src/meho_backplane/runbooks/schemas.py` clean.

## Out of scope

- **Service layer** (create_draft, fork_from_published, etc.) — G12.2-T2.
- **REST routes** — G12.2-T3.
- **MCP tools** — G12.2-T4.
- **Multi-session drafting docs** — G12.2-T5.
- **Substitution at advance time** — G12.3 reuses `validate_substitutions()` to re-check at runtime, but the engine itself is out of scope here.
- **Run-side schemas** (`StartRunRequest`, `NextStepResponse`, etc.) — G12.3 ships its own schemas module under `runbooks/runs_schemas.py` (or alongside; the G12.3 lead decides) since the lifecycle differs. T1 is template-shape only.

## References

- Parent initiative: [#1197](https://github.com/evoila/meho/issues/1197) — G12.2 Template lifecycle.
- Parent goal: [#1195](https://github.com/evoila/meho/issues/1195) — G12 Runbooks.
- Predecessor: [#1292](https://github.com/evoila/meho/issues/1292) — schema migration (provides the underlying tables).
- Design source: [#1191](https://github.com/evoila/meho/issues/1191).
- Slug regex contract being reused: [backend/src/meho_backplane/kb/schemas.py:79](backend/src/meho_backplane/kb/schemas.py#L79).
- Frozen-Pydantic + `__all__` discipline reference: [backend/src/meho_backplane/kb/schemas.py](backend/src/meho_backplane/kb/schemas.py).
- Model-level validator pattern (closest precedent for step-shape validation): [backend/src/meho_backplane/scheduler/schemas.py:103](backend/src/meho_backplane/scheduler/schemas.py#L103).
- Pydantic v2 discriminated unions: https://docs.pydantic.dev/2.10/concepts/unions/#discriminated-unions
- Pydantic v2 `ConfigDict(frozen=True)`: https://docs.pydantic.dev/2.10/concepts/models/#model-config
- **Best practices — Python pack**: Pydantic v2 frozen-by-default for IO models, discriminated unions over runtime-branching, `Annotated[...]` for field-level configuration, validators raise `ValueError` for clean 422 surfacing at HTTP boundary. `../explore-and-refactor/python_best_practices.md` §Pydantic + §Validation.
- **Best practices — AI engineering pack** (substrate minimalism — same call as #1177): no DSL, no expressions, no comparison operators in verify; structural equality + presence match only. `../implement-issue/ai_engineering_best_practices.md` §Determinism over expressivity.

### PR DESCRIPTION
**Pull request by damir-topic**
## Summary
- Lands `backend/src/meho_backplane/runbooks/schemas.py` — the Pydantic v2 shape contract for the G12.2 template lifecycle. Pure types: no service logic, no routes, no tools.
- Step shape is a discriminated union on `type` (`operation_call` | `manual`), each step carrying a `verify` discriminated union on `type` (`confirm` | `operation_call`). The discriminated-union + `model_validator(mode="after")` posture mirrors `ScheduledTriggerCreate` so a malformed body surfaces as a clean 422 at the HTTP boundary.
- `validate_substitutions()` is module-level and exported: it walks a value recursively and rejects every `${...}` pattern except `${run.target}` and `${run.params.X}` (X matching `[a-z_][a-z0-9_]*`). G12.3 reuses it at advance time (defense in depth). The template validator also enforces step-id uniqueness.
- `slug` validation imports `SLUG_PATTERN` from `kb.schemas` verbatim (no redefinition); step ids use the tighter `STEP_ID_PATTERN`. Every model is `ConfigDict(frozen=True)`; `__all__` is alphabetical and complete.

## Test plan
- [x] `ruff check src/meho_backplane/runbooks/ tests/test_runbooks_schemas.py` — clean
- [x] `ruff format --check` — clean (3 files already formatted)
- [x] `mypy src/meho_backplane/runbooks/schemas.py` — clean
- [x] `pytest tests/test_runbooks_schemas.py -v` — 19 passed (covers all 14 specified cases + datetime-bearing summary)
- [x] code-quality hook (`--diff`) — 0 blockers, 1 warning (schemas.py 450 lines, under 600 block threshold)
- [x] Both step types + both verify types parse; unknown `type` tags, missing `id`, bad-pattern `id` all reject
- [x] `${run.target}` / `${run.params.X}` accepted in body / op params / verify params / verify expect; `${run.bad}`, nested `${run.params.X.Y}`, capitalised, and arbitrary patterns rejected
- [x] Duplicate step ids reject; fork-info path + optional filter fields + template-body round-trip verified; `validate_substitutions` importable for G12.3

## Out of scope
- Service layer (T2), REST routes (T3), MCP tools (T4), multi-session drafting docs (T5).
- Substitution at advance time (G12.3 reuses the helper; engine out of scope here).
- Run-side schemas (`StartRunRequest`, etc.) — G12.3 ships its own module.

Closes #1295

## Summary by CodeRabbit

* **New Features**
  * Introduced runbook templates with versioned procedures and step-based execution.
  * Added support for multiple step types with configurable verification gates.
  * Enabled template lifecycle management (draft, publish, deprecate) and editing capabilities.
  * Implemented validation for template structure and parameter substitutions.

* **Tests**
  * Added comprehensive test coverage for runbook schema validation and template operations.

[Review Change Stack](https://app.coderabbit.ai/change-stack/evoila/meho/pull/1331?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by damir-topic** [comment]
@coderabbitai review

File: backend/src/meho_backplane/runbooks/schemas.py [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+    layer (T2) serialises into the ``steps`` JSONB column (alongside the
+    ``title`` / ``description`` / ``target_kind`` columns it lifts out).
+    :attr:`steps` is ordered; each step's verify gates advance to the
+    next at run time.
+
+    The :meth:`_validate_step_ids_unique_and_substitutions_allowlisted`
+    validator enforces the two template-level invariants the per-step
+    models cannot see on their own: step-id uniqueness across the
+    template, and the ``${...}`` substitution allowlist over every
+    string the template carries.
```
Comment (damir-topic) [inline backend/src/meho_backplane/runbooks/schemas.py:247]: **[⚠️ M1]** — Class docstring over-claims the substitution allowlist scope (also flagged by CodeRabbit)

**What:** This docstring says the allowlist runs `over every string the template carries`, but `_validate_step_ids_unique_and_substitutions_allowlisted` walks only `step.body`, `OperationCallStep.params`, `OperationCallVerify.params`, and `OperationCallVerify.expect`. `title`, `description`, and `ConfirmVerify.prompt` are not walked. (The method docstring at L270-272 is precise; this class-level one is not.)

**Why it matters:** The class docstring is the in-code contract for this validator. As written it promises a broader guarantee than the code delivers, which will mislead the G12.2 service (T2) and G12.3 engine authors about which fields are publish-time-clean.

**Required change:** Make docstring and code agree. The minimal, spec-aligned fix is to tighten this line to name the actually-walked surfaces (matching #1295's validator contract: `step.body / step.params / verify.params / verify.expect`):

```suggestion
    template, and the ``${...}`` substitution allowlist over every step
    body, op-call params, verify params, and verify expect the template
    carries (step/template titles, descriptions, and confirm prompts are
    not substitution surfaces and are not walked).
```

Alternatively, if `title` / `description` / `ConfirmVerify.prompt` are intended to be substitution surfaces, widen the validator to walk them too — but that exceeds #1295's stated scope, so the docstring tightening is the in-scope resolution.

**Source:** #1295 validator contract (walks `step.body / step.params / verify.params / verify.expect`); method docstring at schemas.py:270-272.

**Review by damir-topic — COMMENTED** [review]
## /auto-review-pr summary (iteration 1)




**Anchor task:** #1295 — G12.2-T1 Runbook Pydantic schemas + step-shape discriminated-union validation
**Surface area:** backend
**Best-practices packs audited:** python, ai
**Verdict:** APPROVE
**Head SHA reviewed:** `8a71b16`
**AI reviewer (CodeRabbit):** received in ~300s (2 actionable comments)

### Acceptance criteria
- [x] Package `backend/src/meho_backplane/runbooks/` exists with `__init__.py` + `schemas.py` — both present in diff
- [x] `Step` and `Verify` are `Annotated[... | ..., Field(discriminator="type")]` discriminated unions — verified; unknown tags reject (cross-checked against Pydantic 2.10 discriminated-unions docs)
- [x] `RunbookTemplateBody` validates step-id uniqueness — `test_step_ids_must_be_unique` green
- [x] `validate_substitutions()` rejects every `${...}` except `${run.target}` / `${run.params.X}` (X = `[a-z_][a-z0-9_]*`); module-level + exported — verified by an independent adversarial edge-case sweep (nested paths, empty braces, whitespace, caps, dict-key smuggling all fail-closed)
- [x] `STEP_ID_PATTERN` defined; `SLUG_PATTERN` imported from `kb.schemas` (not redefined) — confirmed
- [x] Every model `ConfigDict(frozen=True)` — confirmed on all 14 models
- [x] `__all__` alphabetical + complete — confirmed
- [x] `pytest` green — 19 passed (14 specified + 5 extra)
- [x] `ruff check` + `mypy` clean — both clean in worktree

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #1295 | Open | Closes | ✅ | ✅ full | close on merge |

No sibling PRs open on evoila/meho; no collision risk. No other issues claimed or delivered-but-unlinked.

### Predicted vs. actual diff
- ✅ `backend/src/meho_backplane/runbooks/__init__.py` — package marker + module docstring
- ✅ `backend/src/meho_backplane/runbooks/schemas.py` — the shape contract
- ✅ `backend/tests/test_runbooks_schemas.py` — behavioural coverage
- 🕳️ Missing: none — pure-types scope; service/routes/tools correctly deferred to T2–T5

### Integration verification
- [x] `ruff check src/meho_backplane/runbooks/ tests/test_runbooks_schemas.py` — clean
- [x] `ruff format --check` — 3 files already formatted
- [x] `mypy src/meho_backplane/runbooks/schemas.py` — clean
- [x] `pytest tests/test_runbooks_schemas.py` — 19 passed
- [x] Adversarial substitution-validator sweep run in the review worktree — all edge cases fail-closed
- ⚠️ not run: migrations / docker compose / agent eval — n/a for a pure-types module

### Findings (inline — each is a merge blocker per team policy)
- ⚠️ M1 — Class docstring (schemas.py:247) claims the allowlist covers `every string the template carries`, but the validator walks only `step.body / step.params / verify.params / verify.expect` (not `title` / `description` / `ConfirmVerify.prompt`). Required change: tighten the class docstring to match the spec'd surfaces (the in-scope fix), or widen the validator (exceeds #1295 scope). Does not block the verdict — no acceptance criterion is unmet — but should be resolved before merge.

### Adjacent improvements (non-blocking, not posted inline)
- CodeRabbit's "shallow `frozen=True`" finding (nested `dict`/`list` remain mutable) is **technically true but disputed as a change request**: every sibling schema module #1295 tells the author to mirror — `kb.schemas`, `scheduler.schemas`, `memory.schemas` — uses plain `ConfigDict(frozen=True)` with `dict[str, object]` / `list[...]` fields and none wrap in `MappingProxyType` / `frozendict` / `tuple`. `kb.schemas` even documents this explicitly ("attribute reassignment only -- the metadata dict ... [is mutable]"). The task spec mandates exactly `ConfigDict(frozen=True)`. Adopting deep immutability here would diverge from a deliberate, documented codebase convention and exceed task scope. If deep immutability is wanted, it belongs in a cross-cutting decision applied to all schema modules at once — recorded here for a future initiative, not this PR.

### AI reviewer (CodeRabbit) cross-check
- Adopted: 1 (M1 — docstring/scope mismatch; cited inline)
- Disputed: 1 (shallow-frozen deep-immutability request — contrary to documented codebase convention + task scope; rationale above)
- Out of scope: 0

### Docs
- [x] n/a — no `docs/codebase/runbooks.md` expected for a pure-types T1; sibling feature docs (`kb.md`, `scheduler.md`, `memory.md`) cover full features (service + routes + tools), and the discriminator/shape convention is already documented in `docs/codebase/api-shape-conventions.md`. The feature-level doc belongs with T2–T5 as the surface materialises. Module docstrings here are thorough and serve as the interim contract.

### What worked well (review-body only — never posted as inline comments)
- The `validate_substitutions` helper is genuinely fail-closed: dict keys are walked (key-smuggling defense), nested path / empty-brace / whitespace / caps variants all reject, and the two allowlisted forms are accepted everywhere — verified by an independent adversarial sweep, not just the bundled tests.
- Discriminated-union posture (`Annotated[A | B, Field(discriminator="type")]` with `Literal` tags) is idiomatic Pydantic v2 and mirrors `ScheduledTriggerCreate` as the task asked; unknown tags reject cleanly via `TypeAdapter`.
- Test suite exceeds the 14 specified cases (19 total) and exercises the validator through both the standalone helper and the publish-time template path.
